pinot-controller/src/main/java/org/apache/pinot/controller/api/lineage/DefaultLineageManager.java

```diff
@@ -0,0 +1,125 @@
+
+  public DefaultLineageManager(ControllerConf controllerConf) {
+    _controllerConf = controllerConf;
+  }
+
+  @Override
+  public void updateLineageForStartReplaceSegments(TableConfig tableConfig, String lineageEntryId,
+      Map<String, String> customMap, SegmentLineage lineage) {
+  }
+
+  @Override
+  public void updateLineageForEndReplaceSegments(TableConfig tableConfig, String lineageEntryId,
+      Map<String, String> customMap, SegmentLineage lineage) {
+  }
+
+  @Override
+  public void updateLineageForRevertReplaceSegments(TableConfig tableConfig, String lineageEntryId,
+      Map<String, String> customMap, SegmentLineage lineage) {
+  }
+
+  @Override
+  public void updateLineageForRetention(TableConfig tableConfig, SegmentLineage lineage, List<String> allSegments,
+      List<String> segmentsToDelete, Set<String> consumingSegments) {
+    // 1. The original segments can be deleted once the merged segments are successfully uploaded
+    // 2. The zombie lineage entry & merged segments should be deleted if the segment replacement failed in
+    //    the middle
+    Set<String> segmentsForTable = new HashSet<>(allSegments);
+    Iterator<LineageEntry> lineageEntryIterator = lineage.getLineageEntries().values().iterator();
+    while (lineageEntryIterator.hasNext()) {
+      LineageEntry lineageEntry = lineageEntryIterator.next();
+      if (lineageEntry.getState() == LineageEntryState.COMPLETED) {
+        Set<String> sourceSegments = new HashSet<>(lineageEntry.getSegmentsFrom());
+        sourceSegments.retainAll(segmentsForTable);
+        if (sourceSegments.isEmpty()) {
+          // If the lineage state is 'COMPLETED' and segmentFrom are removed, it is safe clean up the lineage entry
+          lineageEntryIterator.remove();
+        } else {
+          // If the lineage state is 'COMPLETED' and we already preserved the original segments for the required
+          // retention, it is safe to delete all segments from 'segmentsFrom'
+          if (shouldDeleteReplacedSegments(tableConfig, lineageEntry)) {
+            segmentsToDelete.addAll(sourceSegments);
+          }
+        }
+      } else if (lineageEntry.getState() == LineageEntryState.REVERTED || (
+          lineageEntry.getState() == LineageEntryState.IN_PROGRESS && lineageEntry.getTimestamp()
+              < System.currentTimeMillis() - LINEAGE_ENTRY_CLEANUP_RETENTION_IN_MILLIS)) {
+        // If the lineage state is 'IN_PROGRESS' or 'REVERTED', we need to clean up the zombie lineage
+        // entry and its segments
+        Set<String> destinationSegments = new HashSet<>(lineageEntry.getSegmentsTo());
+        destinationSegments.retainAll(segmentsForTable);
+        if (destinationSegments.isEmpty()) {
+          // If the lineage state is 'IN_PROGRESS or REVERTED' and source segments are already removed, it is safe
+          // to clean up the lineage entry. Deleting lineage will allow the task scheduler to re-schedule the source
```

Comment:
Is it source segments or destinations segments in the comments here?

Reply:
This is destination segments, for REVERTED/IN_PROGRESS (more than 24hrs) states, we make sure all segmentsTo are deleted before cleaning up the lineage entries, else both segmentsFrom and segmentsTo are used to answer queries.